File: 0003LongestSubstringWithoutRepeatingCharacters/Main.java
import java.util.HashMap;

public class Main {

    public static int lengthOfLongestSubstring(String s) {
        int i = -1, j = 0, len = 0;
        HashMap<Character, Integer> letters = new HashMap<>();
        while(j < s.length()) {
            if(letters.containsKey(s.charAt(j))) {
                i = Math.max(i, letters.get(s.charAt(j)));
            }
            letters.put(s.charAt(j), j);
            len = Math.max(j - i, len);
            j++;
        }
        return len;
    }

    public static void main(String[] args) {
        System.out.println(lengthOfLongestSubstring("tmmzuxt")); // 5
        System.out.println(lengthOfLongestSubstring("abcabcbb")); // 3
        System.out.println(lengthOfLongestSubstring("bbbbb")); // 1
        System.out.println(lengthOfLongestSubstring("pwwkew")); // 3
        System.out.println(lengthOfLongestSubstring("abbcaccabdac")); // 4
    }
}

File: 0003LongestSubstringWithoutRepeatingCharacters/main.cpp
#include <cstdio>
#include <string>
#include <set>    

using namespace std;

int lengthOfLongestSubstring(string s) {
    int len = 0;
    std::set<int> track;
    int i = 0, j = 0;

    while(j < s.length()) {
        while(track.find(s[j]) != track.end()) {
            track.erase(s[i]);
            i++;
        }
        track.insert(s[j]);
        len = max(len, j - i + 1);

        j++;
    }

    return len;
}

int main() {
    printf("%d\n", lengthOfLongestSubstring("abcabcbb"));
    printf("%d\n", lengthOfLongestSubstring("bbbbb"));
    printf("%d\n", lengthOfLongestSubstring("pwwkew"));
    return 0;
}